diff -urN util-linux-ng-2.13.org/sys-utils/Makefile.am util-linux-ng-2.13/sys-utils/Makefile.am
--- util-linux-ng-2.13.org/sys-utils/Makefile.am	2007-08-27 20:42:22.000000000 +0200
+++ util-linux-ng-2.13/sys-utils/Makefile.am	2007-08-28 09:05:34.716475853 +0200
@@ -76,3 +76,7 @@
 	for I in $(SETARCH_LINKS); do \
 		cd $(DESTDIR)$(usrbinexecdir) && ln -sf setarch $$I ; \
 	done
+	for I in $(SETARCH_LINKS); do \
+                echo ".so setarch.8" > $(DESTDIR)$(man8dir)/$$I.8 ; \
+        done
+
